src/code-util.jl

```diff
@@ -123,25 +123,55 @@ function is_special_symbol(symbol::Symbol)::Bool
     return symbol == :abort || symbol == :output || symbol == :rel
 end
 
-# Generate a string representing the Rel type for the input
-# Expected inputs are a vector of types, a tuple of types, or a type
-function type_string(input::Vector)
-    if isempty(input)
-        return ""
+# Generate a string representing the Rel type for input Vectors
+# :a => [...]
+function type_string(input::Vector{T}) where T
+    # []
+    T == Any && return ""
+    # Non-container type
+    T == eltype(T) && return "/$T"
+    # Strings (otherwise treated as containers by Julia)
+    T == String && return "/$T"
+
+    # Container type
+    result = ""
+    for e_type in fieldtypes(T)
+        result *= type_string(e_type)
     end
-    return type_string(input[1])
+
+    return result
 end
 
-function type_string(input::Tuple)
+# Generate a string representing the Rel type for input Tuples
+# :a => (...)
+function type_string(input::T) where { T <: Tuple}
     result = ""
-    for t in input
-        result *= type_string(t)
+    for e_type in fieldtypes(T)
+        result *= type_string(e_type)
     end
+
     return result
 end
 
-function type_string(input)
-    return "/" * string(typeof(input))
+# Generate a string representing the Rel type for input values
+# :a => 1
+type_string(input::T) where T = "/" * string(T)
+
+# Generate a string representing the Rel type for input types
+# We've extracted the type from the input so now we can convert it to a String result
+function type_string(type::Type{T}) where T
+    # Strings (otherwise treated as containers by Julia)
+    type == String && return "/$type"
```

Comment:
this method looks half updated (`test` vs `T`), but the tests pass - could we add some test coverage here?

Reply:
ah, wait - I see it now. `type == T` - could we just use `T` here?

Reply:
We can and now we have. I've simplified this further, so much that I think it needs re-approval to check it still makes sense.